sql/logic_test.go

```diff
@@ -201,22 +199,21 @@ func (t *logicTest) setUser(user string) func() {
 	}
 	if db, ok := t.clients[user]; ok {
 		t.db = db
-		// No cleanup necessary, but return a no-op func to avoid nil pointer dereference.
-		return func() {}
+		return
 	}
 
-	credsDir, err := ioutil.TempDir("", "user-credentials")
+	credsDir, err := ioutil.TempDir(tempDir, "user-credentials-"+user)
 	if err != nil {
 		t.Fatal(err)
 	}
-	pgUrl, cleanupFunc := sqlutils.PGUrl(t.T, &t.srv.TestServer, user, credsDir, "TestLogic")
+	pgUrl, _ := sqlutils.PGUrl(t.T, &t.srv.TestServer, user, credsDir, "TestLogic")
```

Comment:
If we're going to discard the `cleanupFunc` we should update the `PGUrl` docs to say that the `cleanupFunc` need not be called if the caller deletes the entire `tempDir` that is passed in.

In fact, all callers of `PGUrl` that do not do this are currently broken; these other calls are responsible for #4107. Most calls to `PGUrl` pass `os.TempDir()` (i.e. `/tmp`) and `PGUrl` reuses the same filenames, so every test run is creating and deleting the same files and occasionally they hit a race. We need to either remove the `tempDir` argument from `PGUrl` and have it construct a new unique temporary directory, or require all callers to pass in a unique directory and make the caller responsible for deleting it (eliminating the `cleanupFunc` result).